### PR TITLE
*: Provide `TemporaryTableSnapshotInterceptor` to read temporary table data

### LINKED ISSUE #28035: Implement snapshot interceptor for temporary table

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #28035

Related to #27936 , in the previous pr we provide interfae `SnapshotInterceptor` to interceptor snapshot read operations. This pr provides a implement for `SnapshotInterceptor` to read temporary table data.

### What is changed and how it works?

Provide `TemporaryTableSnapshotInterceptor` to read temporary table data

### Check List 

Tests 

- [x] Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by djshow832 — COMMENTED** [review]
Rest LGTM

File: table/temptable/interceptor.go [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+	if bytes.HasPrefix(k, tablecodec.TablePrefix()) && len(k) >= tableWithIDPrefixLen {
+		if tbID := tablecodec.DecodeTableID(k); tbID > 0 && tbID != math.MaxInt64 {
+			return tbID, true
+		}
+	}
+
+	return 0, false
+}
+
+func safeWithSnapshotScan(k, upperBound kv.Key) bool {
```
Comment (tiancaiamao) [inline table/temptable/interceptor.go:253]: Is `notTableRange` a better name?

**Comment by tiancaiamao** [comment]
Only 250 lines is the actual code, and more than 1700 lines is testing...
This is really a very good habit!

File: table/temptable/ddl_test.go [hunk 2]
```diff
@@ -27,16 +27,9 @@ import (
 	"github.com/pingcap/tidb/tablecodec"
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/util/mock"
-	"github.com/pingcap/tidb/util/testbridge"
 	"github.com/stretchr/testify/require"
-	"go.uber.org/goleak"
 )
 
-func TestMain(m *testing.M) {
```
[the remark this answers is not in the document]
Comment (lcwangchao) [reply table/temptable/ddl_test.go:35]: It's not removed. `TestMain` is moved to `main_test.go` which provides some common methods for test.

File: table/temptable/interceptor.go [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/store/driver/txn"
+	"github.com/pingcap/tidb/tablecodec"
+)
+
+var (
```
[the remark this answers is not in the document]
Comment (lcwangchao) [reply table/temptable/interceptor.go:30]: Const is better, however golang does not support initializing a const with some methods like `len`

File: table/temptable/interceptor.go (hunk 1 above)
Comment (lcwangchao) [reply table/temptable/interceptor.go:253]: OK. It's too hard for me to choose with name is better...

**Review by djshow832 — APPROVED** [review]

**Review by tiancaiamao — APPROVED** [review]

**Pull request merged**